src/main/java/nextstep/security/authorization/hierarchy/RoleHierarchyImpl.java

```diff
@@ -0,0 +1,93 @@
+package nextstep.security.authorization.hierarchy;
+
+import org.springframework.util.Assert;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+
+public class RoleHierarchyImpl implements RoleHierarchy {
+    private final Map<String, Set<String>> rolesReachableInOneOrMoreStepsMap;
+
+    private RoleHierarchyImpl(final Map<String, Set<String>> rolesReachableInOneOrMoreStepsMap) {
+        this.rolesReachableInOneOrMoreStepsMap = rolesReachableInOneOrMoreStepsMap;
+    }
+
+    public static Builder with() {
+        return new Builder();
+    }
+
+    @Override
+    public Collection<String> getReachableAuthorities(Collection<String> authorities) {
+        if (authorities == null || authorities.isEmpty()) {
+            return new ArrayList<>();
+        }
+
+        Set<String> reachableRoles = new HashSet<>();
+        Set<String> processedRoles = new HashSet<>();
+        Queue<String> queue = new LinkedList<>(authorities);
```

Comment:
Spring Security는 성능 최적화를 중요하게 생각하는데요,
ArrayDeque 사용을 통해 성능을 향상 시켜봐도 좋을 것 같습니다 😄 
```suggestion
        Queue<String> queue = new ArrayDeque<>(authorities);
```